### PR TITLE
Make s3 repository sensitive settings use secure settings

### PR DESCRIPTION
**Pull request by rjernst**
This change converts repository-s3 to use the new secure settings. In
order to support the multiple ways we allow aws creds to be configured,
it also moves the main methods for the keystore wrapper into a
SecureSettings interface, in order to allow settings prefixing to work.

### CODE REVIEW

**Comment by rjernst** [comment]
I still need to figure out some tests for this, but I wanted to get this up here to get some feedback on how it looks.

File: core/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java [hunk 1]
```diff
@@ -225,32 +226,41 @@ void save(Path configDir) throws Exception {
...
+    }
+
     // TODO: make settings accessible only to code that registered the setting
     /** Retrieve a string setting. The {@link SecureString} should be closed once it is used. */
-    SecureString getStringSetting(String setting) throws GeneralSecurityException {
-        KeyStore.Entry entry = keystore.get().getEntry(setting, keystorePassword.get());
-        if (entry instanceof KeyStore.SecretKeyEntry == false) {
-            throw new IllegalStateException("Secret setting " + setting + " is not a string");
+    @Override
+    public SecureString getString(String setting) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java]: can't we throw `GeneralSecurityException` do we have to wrap?

**Review by s1monw — APPROVED** [review]
LGTM I don't really know how we should test this. I think the simplest would be to have a class that you can override that adds factory methods for `StaticCredentialsProvider` etc. instead of using a static method.?

File: core/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java (hunk 1 above)
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java]: Sure, we could throw it. I was thinking ahead to Vault integration, but I guess that could make GeneralSecurityExceptions?

Comment (rjernst) [inline core/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java]: I take it back. Unfortunately, GeneralSecureException is not the only thing that can be thrown. Not all keystore apis throw a subclass of this. There are like 3 different exceptions. :(

Comment (rjernst) [inline core/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java]: Oops, nevermind. Done!

**Comment by rjernst** [comment]
@s1monw I found the existing tests for the old settings, made them work with deprecation log checks, and duplicated them to use new secure settings.

**Comment by rjernst** [comment]
retest this please

**Review by s1monw — APPROVED** [review]
LGTM

**Comment by rjernst** [comment]
retest this please

**Comment by dadoonet** [comment]
@rjernst I tried some first tests today:

# With elasticsearch 5.1.1

In my `config/elasticsearch.yml` file:

```yml
cloud.aws.access_key: "ACCESS-KEY"
cloud.aws.secret_key: "SECRET-KEY"
repositories.s3.bucket: "test.eu-west-1.elasticsearch.org"
repositories.s3.region: "eu-west-1"
```

Then I start elasticsearch and run:

```sh
curl -XPOST 'http://localhost:9200/_snapshot/backups?pretty=true' -d '{ "type":"s3" }'
```

This gives:

```json
{
  "acknowledged" : true
}
```

# With elasticsearch 6.0.0

```sh
bin/elasticsearch-keystore create
bin/elasticsearch-keystore add cloud.aws.secret_key
# Entering here ACCESS-KEY
bin/elasticsearch-keystore add cloud.aws.access_key
# Entering here SECRET-KEY
```

In `elasticsearch.yml`:

```yml
repositories.s3.bucket: "test.eu-west-1.elasticsearch.org"
repositories.s3.region: "eu-west-1"
```

Then I start elasticsearch and run:

```sh
curl -XPOST 'http://localhost:9200/_snapshot/backups?pretty=true' -d '{ "type":"s3" }'
```

This gives:

```json
{
  "error" : {
    "root_cause" : [
      {
        "type" : "repository_verification_exception",
        "reason" : "[backups] path  is not accessible on master node"
      }
    ],
    "type" : "repository_verification_exception",
    "reason" : "[backups] path  is not accessible on master node",
    "caused_by" : {
      "type" : "i_o_exception",
      "reason" : "Unable to upload object tests-Z8dZ3RULTSKNXfoVU222Rw/master.dat-temp",
      "caused_by" : {
        "type" : "amazon_s3_exception",
        "reason" : "The AWS Access Key Id you provided does not exist in our records. (Service: Amazon S3; Status Code: 403; Error Code: InvalidAccessKeyId; Request ID: FE35B1FCEC30375A)"
      }
    }
  },
  "status" : 500
}
```

May be I forgot something though.
I'm going to look at it in debug mode and will update here with findings.

**Comment by dadoonet** [comment]
Apparently I'm getting back an empty String when running https://github.com/elastic/elasticsearch/blob/8015fbbf257c7e55210f13cb5dbc311fcf9feaf1/plugins/repository-s3/src/main/java/org/elasticsearch/cloud/aws/InternalAwsS3Service.java#L131-L131

This is then trying to use environment variables, system properties or instance profile credentials. Then it causes that error message.

I don't understand why I'm getting an empty String while asking for `repositories.s3.access_key` and `repositories.s3.secret_key`.
Note that I also added those two settings with `elasticsearch-keystore` app:

```sh
$ bin/elasticsearch-keystore list
cloud.aws.access_key
cloud.aws.secret_key
repositories.s3.access_key
repositories.s3.secret_key
```

**Pull request merged**